src/psyclone/psyir/backend/base.py

```diff
@@ -0,0 +1,181 @@
+            raise TypeError(
+                "skip_nodes should be a boolean but found '{0}'."
+                "".format(type(skip_nodes).__name__))
+        if indent_string is not None and not isinstance(indent_string, str):
+            raise TypeError(
+                "indent_string should be a str but found '{0}'."
+                "".format(type(indent_string).__name__))
+        if not isinstance(initial_indent_depth, int):
+            raise TypeError(
+                "initial_indent_depth should be an integer but found '{0}'."
+                "".format(type(initial_indent_depth).__name__))
+        if initial_indent_depth < 0:
+            raise TypeError(
+                "initial_indent_depth should not be negative, but found '{0}'."
+                "".format(initial_indent_depth))
+
+        self._skip_nodes = skip_nodes
+        if indent_string is None:
+            self._indent = "  "
+        else:
+            self._indent = indent_string
+        self._depth = initial_indent_depth
+
+    @property
+    def _nindent(self):
+        '''
+        :returns: the current indentation string.
+        :rtype: str
+
+        '''
+        return self._depth * self._indent
+
+    def __call__(self, node):
+        '''This method is called when an instance of the class is called
+        directly (like a function). This implementation is known as
+        a functor. It makes sense for this class as there is only one
+        main method - the `visit` method.
+
+        :param node: A PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Node`
+
+        '''
+        return self.visit(node)
+
+    def visit(self, node):
+        '''Implements the PSyIR callbacks. Callbacks are implemented by using
+        the class hierarchy names of the object in the PSyIR tree as
+        the candidate method names. The class name of the object is
+        tried first, then the class name of its parent, and so on
+        until there are no more parent classes. Names are not
+        modified, other than making them lower case, apart from the
+        `Return` class which is changed to `return_node` because
+        `return` is a Python keyword.
+
+        :param node: A PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Node`
+
+        :raises VisitorError: if a node is found that does not have \
+        associated call back methods (and skip_nodes is not set).
+        :raises AttributeError: if a call back method is found but it \
+        raises an AttributeError.
+
+        '''
+        if not isinstance(node, Node):
+            raise VisitorError(
+                "Expected argument to be of type 'Node' but found '{0}'."
+                "".format(type(node).__name__))
+
+        # Make a list of the node's ancestor classes (including
+        # itself) in method resolution order (mro).
+        import inspect
+        possible_method_names = [curr_class.__name__.lower() for curr_class
```

Comment:
We probably should remove the 'object' class from the possible_method_names.

Reply:
I don't think it's a big deal. It could even have been a feature, e.g. one could raise an exception if the node method was ever called in case a node in PSyIR did not subclass `Node`. However, I have removed it as requested.